k8sclient/k8sclient.go

```diff
@@ -326,30 +331,60 @@ func cniConfigFromNetworkResource(customResource *types.NetworkAttachmentDefinit
 	return config, nil
 }
 
-func getKubernetesDelegate(client KubeClient, net *types.NetworkSelectionElement, confdir string) (*types.DelegateNetConf, error) {
+func getKubernetesDelegate(client KubeClient, net *types.NetworkSelectionElement, confdir string, podID string, resourceMap map[string]*types.ResourceInfo) (*types.DelegateNetConf, map[string]*types.ResourceInfo, error) {
+
 	logging.Debugf("getKubernetesDelegate: %v, %v, %s", client, net, confdir)
 	rawPath := fmt.Sprintf("/apis/k8s.cni.cncf.io/v1/namespaces/%s/network-attachment-definitions/%s", net.Namespace, net.Name)
 	netData, err := client.GetRawWithPath(rawPath)
 	if err != nil {
-		return nil, logging.Errorf("getKubernetesDelegate: failed to get network resource, refer Multus README.md for the usage guide: %v", err)
+		return nil, resourceMap, logging.Errorf("getKubernetesDelegate: failed to get network resource, refer Multus README.md for the usage guide: %v", err)
 	}
 
 	customResource := &types.NetworkAttachmentDefinition{}
 	if err := json.Unmarshal(netData, customResource); err != nil {
-		return nil, logging.Errorf("getKubernetesDelegate: failed to get the netplugin data: %v", err)
+		return nil, resourceMap, logging.Errorf("getKubernetesDelegate: failed to get the netplugin data: %v", err)
+	}
+
+	// Get resourceName annotation from NetDefinition
+	deviceID := ""
+	resourceName, ok := customResource.Metadata.Annotations[resourceNameAnnot]
+	if ok && podID != "" {
+		// ResourceName annotation is found; try to get device info from resourceMap
+		logging.Debugf("getKubernetesDelegate: found resourceName annotation : %s", resourceName)
+
+		if resourceMap == nil {
+			checkpoint, err := checkpoint.GetCheckpoint()
+			if err != nil {
+				return nil, resourceMap, logging.Errorf("getKubernetesDelegate: failed to get a checkpoint instance: %v", err)
+			}
+			resourceMap, err = checkpoint.GetComputeDeviceMap(podID)
+			if err != nil {
+				return nil, resourceMap, logging.Errorf("getKubernetesDelegate: failed to get resourceMap from kubelet checkpoint file: %v", err)
+			}
+			logging.Debugf("getKubernetesDelegate(): resourceMap instance: %+v", resourceMap)
+		}
+
+		entry, ok := resourceMap[resourceName]
+		if ok {
+			if idCount := len(entry.DeviceIDs); idCount > 0 && idCount > entry.Index {
```

Comment:
> one q. why do you add 'idCount > 0'?
> I suppose you may want to add 'entry.Index >=0', as entry.Index should be between 0 and len(entry.DeviceIDs).

DeviceIDs array may have '0' elements in it, so we need to ensure there exists enough element(s) before we try to get a deviceID with an index value from that array, just to be safe. The entry.Index will index the deviceID we need for this  delegate object. Let me clear this up with an example:

resouceMap["intel.com/sriov_net_A"] returns following ResourceInfo object

```
{
    Index: 0,
    DeviceIDs: [ "0000:03:0e.1", "0000:03:0e.3"]
}
```

We want to ensure first delegate with resourceAnnotation "intel.com/sriov_net_A" will get DeviceIDs[Index] - "0000:03:0e.1". Then the Index will be incremented ( i.e 1 ) as we have consumed one deviceID for this delegate. When another delegate reference the same resourceAnnotation will get DeviceIDs[1] - "0000:03:0e.3". The checks in this code block ensure that we have enough elements in the DeviceIDs array that we are about to index. Because after the 2nd delegate, the Index will be 2 but we do not have any more deviceIDs left to attach. A CRD still may reference a resourceAnnotation with no resource request ever be made in Pod spec with correct reference.

I hope that answers your question.

Reply:
Answer to Q1:
Kubelet manages resource Allocation for all Pods and updates the checkpoint file. Multus just needs to read the checkpoint file if a CRD has a resourceName annotation.  So, on Pod deletion Kubelet will de-allocate all resources consumed by that Pod and make them available for others.

Q2:
This is a decision we need to make. We can either log the error and continue without device information(ignoring unknown resource annotations) OR return a hard error if CRD contains a resourceAnnotation but no associated device allocation information is found. Let me know which  is preferred.

Reply:
@s1061123 @ahalim-intel Good idea to track the Q2 in the github issue. We should discuss on it..